app/Http/Controllers/QuizController.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Http\Requests\StoreQuizResultRequest;
+use App\Models\Course;
+use Illuminate\Contracts\View\View;
+use Illuminate\Http\RedirectResponse;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Validator;
+
+class QuizController extends Controller
+{
+    public function show(Course $course): View
+    {
+        $quiz = $course->quiz;
+
+        return view('quizzes.show', [
+            'course' => $course,
+            'quiz' => $quiz,
+            'title' => $quiz->title,
+            'questions' => $quiz->questions,
+        ]);
+    }
+
+    public function storeQuizResult(StoreQuizResultRequest $request, Course $course): View | RedirectResponse
+    {
+        $data = $request->validated();
+        $user = Auth::user();
+        $correctQuestions = 0;
+        $quiz = $course->quiz;
+        $quizWithCounts = $quiz->loadCount('questions');
+        $isPass = false;
+        $validator = Validator::make([], []);
+        $previousAnswers = [];
+        $wrongAnswers = [];
+
+        foreach ($quizWithCounts->questions as $question) {
+            $previousAnswers[$question->id] = $data['questions'][$question->id];
+            if ($question->getCorrectChoices() == $data['questions'][$question->id]) {
+                $correctQuestions++;
+            } else {
+                $validator->errors()->add('questions.'.$question->id, __('Wrong answer'));
+                $wrongAnswers[] = $question->id;
+            }
+        }
+
+        //@phpstan-ignore-next-line
+        if ($quizWithCounts->questions_count > 0) {
+            $quizScore = $correctQuestions / $quizWithCounts->questions_count;
+            $quizUser = $quiz->users->find($user->id)?->getRelationValue('pivot');
+            if ($quizUser) {
+                $attempts = $quizUser->attempts;
+                $user->quizzes()->updateExistingPivot(
+                    $quiz->id, [
+                        'attempts' => $attempts + 1,
+                        'score' => $quizScore,
+                    ]
+                );
+            } else {
+                $user->quizzes()->attach(
+                    $quiz->id, [
+                        'attempts' => 1,
+                        'score' => $quizScore,
+                    ]
+                );
+            }
+            if ($quizScore >= $quiz->minimum_score) {
+                $isPass = true;
+                $user->courses()->updateExistingPivot(
+                    $quiz->course->id, [
+                        'received_certificate_at' => now(),
+                    ]
+                );
+            } else {
+                return redirect(localized_route('quizzes.show', $course))->with(['previousAnswers' => $previousAnswers, 'wrongAnswers' => $wrongAnswers])->withErrors($validator);
+            }
+        }
+
+        return view('quizzes.store-result', ['quiz' => $quiz, 'results' => $isPass]);
```

Comment:
I'm was playing around with the training. It occurs to me that because we allow the quiz to pass with less than 100% correct, users won't know that they've made mistakes. We may always want to redirect back to the test to show the results of their answers if they pass but in those cases show what the correct answers should be. I think we may need to talk to @michelled and @cherylhjli about this.

Reply:
I set the pass threshold to be 100%, so unless users get perfect score on the quiz, they won't get the certificate and see the pass page.